helix: Add match operator

fantacell opened the pull request:
This is an implementation of matching like "m i (", as well as "] (" and "[ (" in `helix_mode` with a few supported objects and a basis for more.

Release Notes:

- Added helix operators for selecting text objects

fantacell committed:
Add a basic helix match operator

fantacell committed:
Solve an inconsistency between vim and helix

fantacell committed:
Add a test

fantacell committed:
Start a helix directory

fantacell committed:
Change how the match operator is displayed

fantacell committed:
Add select.rs to the new helix directory

fantacell committed:
Add select.rs to git tracking

fantacell commented on assets/keymaps/vim.json:
```diff
@@ -568,6 +566,12 @@
     }
   },
+  {
+    "context": "vim_operator == ℳ",
```
Thanks, I went with the second option.

ConradIrwin commented:
Looks like the test is failing on the merge of main and this branch, but passes on both branches separately. Let me know if you need help debugging why..

fantacell committed:
Merge branch 'main' into helix-match-fix to fix the tests

fantacell committed:
Fix marks in vim normal mode

ConradIrwin commented:
@fantacell should this still be a draft?

fantacell committed:
Add the ] and [ operators to helix

fantacell commented:
I didn't know when writing this that helix itself doesn't have motions like selecting the next brackets `](` ([yet?](https://github.com/helix-editor/helix/pull/11695)), but I think they can still be useful.

fantacell committed:
Fix issues:
- fix the operators' display in the status bar
- fill the subword object placeholder
- fix some spelling mistakes
- fix a bug where the first and last words in a buffer couldn't be selected

fantacell committed:
Resolve upstream merge conflicts

fantacell committed:
Fix nested brackets and the failing tests

fantacell committed:
Create a different object operator rather than pushing two operators on one input

fantacell commented:
Right now typing something like `["` quickly will start to select a register, rather than selecting the previous `"` pair. I have to wait until the status bar shows `[` after pressing `[` and I don't really know how to fix that.

fantacell committed:
Rewrite the logic to use multiple traits instead of a lot of methods on
one object. This makes return types shorter and makes it easy to add an
extra helix object entirely. Also add logic for the sentence and
paragraph text objects. The way this is implemented would make it easy
to for example ignore brackets after a backslash.

fantacell committed:
Merge branch 'main' into helix-match-upstream

fantacell committed:
Adjust to upstream changes

ConradIrwin commented:
@fantacell do you have some time next week to pair on this with me? I'd like to understand the keybinding conflict you're seeing more (but otherwise this seems good). https://cal.com/conradirwin/pairing

fantacell committed:
Don't let sentences span across paragraphs

fantacell committed:
Merge branch 'main' into helix-match-upstream-3

fantacell committed:
Fix clippy in boundary.rs

ConradIrwin committed:
Fix helix bindings

fantacell committed:
Move a few more keybindings around

fantacell committed:
Make matching vim objects in helix mode fallible

fantacell committed:
Switch the cursor side of the selection after a [ motion

fantacell committed:
Count the "around range" of an object as part of the object, unless it is only whitespace

fantacell committed:
Mention the differences to helix's way of doing this in the docs

fantacell committed:
Fix a bug where the next fuzzy boundary is not actually the closest one

fantacell committed:
Throw a warning instead of crashing

fantacell committed:
Make it way faster by removing redundant checks

fantacell committed:
Now it can be written more nicely

fantacell committed:
Fix clippy

fantacell committed:
Actually fix clippy

fantacell commented:
It feels ok to me in release builds. But I can store more state in between searches to avoid backtracking, that would be needed for "m i m" anyway. I could do that and switch to only offsets, either now in this PR or later. If it is ok for now, the only problem left is the keybinding conflict.

fantacell committed:
Only calculate the logic with offsets

fantacell committed:
Merge branch 'main' into helix-match-upstream

fantacell commented:
*Much* better. `m i p` is still slower than `v i p`, but more comparable. I only removed all the unnecessary conversions between `DisplayPoint` and offset.
I think this is ready when the keybinding conflict is solved, but I have no idea how to do that.

fantacell commented:
> hey @fantacell great job with this PR. Although it looks like a lot of boundary-related logic got duplicated, if you have @ConradIrwin blessing for this - it probably is the correct way to go. Lovely to see helix-related operators being implemented thoroughly.

thanks! About the duplication: In vim mode every object kind handles the finding itself, which is hard to access to then find the "]" and "[" ranges. In this PR, the only object specific thing is what is and isn't a start / end depending on the surrounding chars. The finding itself is then only done by the "]", "[", "mi" and "ma" operators. Also, "vi" and "mi" ranges are in many cases slightly different and I didn't want to add tons of checks for mode to then cover every edge case. An advantage of this would be that tree-sitter text objects should be easy to add, as well as kakoune style "]" and "[", which is how helix "]p" and "[p" seem to actually work. I just haven't found a way to express that with the keymap. If you (or anyone) have an idea for that, it would be welcome.

> I'd be interested to look into how we can integrate with #34798 next.

How to access this over the keymap does seem like a real problem right now.

ConradIrwin committed:
Fix [{ precedence in operator mode

ConradIrwin commented:
Sorry for the delay here, I've been wrapped up in https://agentclientprotocol.com for the last few weeks.

I think the problem is just we were missing the `operator` vim binding mode in the keymap; pushed a fix.  The helix test is still failing though...

Feeling close though :D.

@romanisch - re duplication; I think it's a better structure overall if helix matching is "consistently slightly different" to not try and share too much code (because it ends up with a bunch of nested booleans and ifs in the middle of stuff that's already quite fiddly). So it seems clearer to have two methods that do the similar-but-not-quite-the-same things.

(In v0 of helix mode, re-using vim code made more sense because the differences didn't matter)

fantacell committed:
Fix the failing test

fantacell commented:
Yes; thank you for fixing the keybinding stuff, it was nice to see green for once. But I found a bug with `m i "` that starts a few commits ago. It shouldn't be hard to fix, but I need a little more time.
Then there is the `[ p` which doesn't work like the other objects in helix, so what it does now is kind of wrong. But maybe that's ok for now?

fantacell committed:
Fix bug with objects with ambiguous ranges

romaninsh commented:
@fantacell, I have combined some of my previous work with the now closed Select mode PR into this draft:

https://github.com/zed-industries/zed/pull/37748

I found it quite hard to figure out how to operate helix_move_cursor in select move, since perform collapse of selection. It's not longer possible to wrap it as i've done before.

Do you have some idea how to implement it without regressions?
